src/sink/mod.rs

```diff
@@ -0,0 +1,298 @@
+    /// additional I/O activity, and in the asynchronous I/O setting, that work
+    /// cannot be done on drop.
+    ///
+    /// This method requests that the sink initiate closing; as with sending an
+    /// item, however, the actual work involved must be processed by calling
+    /// `poll_complete`.
+    ///
+    /// The method is guaranteed to succeed, and should not panic, even if a
+    /// closure request has already occurred.
+    fn start_close_sink(&mut self);
+
+    /// Make progress on all pending requests, and determine whether they have
+    /// completed.
+    ///
+    /// Since sinks are asynchronous, no single method completes all of their
+    /// work in one shot. Instead, you use `poll_complete` to repeatedly drive
+    /// the sink to make progress on requests (such as `start_send` and
+    /// `start_close_sink`). As with `Future::poll`, if the pending requests are
+    /// not able to complete during this call, the current task is automatically
+    /// scheduled to be woken up again once more progress is possible.
+    ///
+    /// # Return value
+    ///
+    /// Returns `Ok(Async::Ready(()))` when no unprocessed requests remain.
+    ///
+    /// Returns `Ok(Async::NotReady)` if there is more work left to do, in which
+    /// case the current task is scheduled to wake up when more progress may be
+    /// possible.
+    ///
+    /// # Errors
+    ///
+    /// Returns `Err` if the sink encounters an error while processing one of
+    /// its pending requests. Due to the buffered nature of requests, it is not
+    /// generally possible to correlate the error with a particular request. As
+    /// with `start_send`, these errors are generally "fatal" for continued use
+    /// of the sink.
+    ///
+    /// # Panics
+    ///
+    /// This method may panic in a few situations, depending on the specific sink:
+    ///
+    /// - It is called outside of the context of a task.
+    /// - A previous call to `start_send` or `poll_complete` yielded a permanent
+    /// error.
+    fn poll_complete(&mut self) -> Poll<(), Self::SinkError>;
+
+    /// Composes a function *in front of* the sink.
+    ///
+    /// This adapter produces a new sink that passes each value through the
+    /// given function `f` before sending it to `self`.
+    ///
+    /// To process each value, `f` produces a *future*, which is then polled to
+    /// completion before passing its result down to the underlying sink. If the
+    /// future produces an error, that error is returned by the new sink.
+    ///
+    /// Note that this function consumes the given sink, returning a wrapped
+    /// version, much like `Iterator::map`.
+    fn with<U, F, Fut>(self, f: F) -> With<Self, U, F, Fut>
+        where F: FnMut(U) -> Fut,
+              Fut: IntoFuture<Item = Self::SinkItem, Error = Self::SinkError>,
+              Self: Sized
+    {
+        with::new(self, f)
+    }
+
+    /*
+    fn with_map<U, F>(self, f: F) -> WithMap<Self, U, F>
```

Comment:
These are just placeholders for additional combinators, which I wanted to sketch without fully implementing just yet.